### PR TITLE
Add support for -H compiler flag

### LINKED ISSUE #2733: Debugging build failures with implicit_transitive_deps false
I'm wondering if there is standard advice on debugging build failures introduced by setting `implicit_transitive_deps false`? For example, I added that config to a project that includes a ppx rewriter and see:
```
Error: The module Selected_ast is an alias for module Ppxlib_ast__.Import.Selected_ast, which is missing
```
I suppose that by digging through the metadata files for ppxlib I would eventually find some library name to add to the stanza:
```
(library
 (name ppx_trace)
 (kind ppx_rewriter)
 (preprocess no_preprocessing)
 (libraries ppxlib))
```
But is there a better message that dune can give, since without disabling `implicit_transitive_deps` the build succeeds?

### LINKED ISSUE #4963: ignore ppx runtime dependencies when implicit transitive deps = false
## Desired Behavior

When `(implicit_transitive_deps false)` is specified, it would be good if dependencies required by ppx rewriters were excluded from it. For example, `ppx_deriving.runtime` should be left implicit even in this case, because in all likelihood it's only depended upon by automatically generated code, for which the dependency is already explicit in `(pps …)`.

related: #2733

### LINKED ISSUE #9333: Add support for the new -H <dir> argument in OCaml 5.2
Would fix:
* https://github.com/ocaml/dune/issues/2733
* https://github.com/ocaml/dune/issues/4963

https://github.com/ocaml/ocaml/pull/12246 added support for the `-H <dir>` argument, which, similarly to `-I <dir>` adds the given directory to the list of directories the compiler will look into for interfaces. However contrary to `-I`, `-H` forbids explicit calls to the modules and instead only resolves their type.

Using this would fix a foundamental flaw in the `(implicit_transitive_deps false)` option which causes regular annoyance for users and opam-repository maintainers alike.

Dune should be able to detect which version of the OCaml compiler it’s using, and if it’s OCaml >= 5.2, then instead of not adding any of the implicit directories in `(implicit_transitive_deps false)` mode, it would instead use `-H`, the same way `-I` is used with `(implicit_transitive_deps true)`

### PR DESCRIPTION
**Pull request by nojb**
This PR adds support to Dune to use the `-H` compiler flag introduced in https://github.com/ocaml/ocaml/pull/12246, which provides good semantics to the `(implicit_transitive_deps false)`.

This work was carried out by @MA0010 as part of an internship at LexiFi. I am opening this PR but @MA0010 will take the lead during the review process.

- The first commit contains the meat of the change: we keep track of which dependencies are "direct" (= mentioned in the Dune file) and which dependencies are purely transitive. For direct dependencies we use `-I` and for transitive dependencies we use `-H` (of course, only for OCaml >= 5.2). The use of `-H` is also guarded by Dune version 3.17 (this is to minimize the possibility of breakage, also there are some corner cases where the change is not completely backwards-compatible). 

- The second commit adds an explicit `unix` dependency to some vendored libraries. This is needed because `unix` is mentioned in the source of these libraries, but the library is not mentioned in the Dune stanzas. The libraries are able to build before this PR because OCaml adds `-I +unix` implicitly (for recent enough versions of OCaml where `unix` is installed in the separate directory `+unix`). However, after this PR, the compiler passes `-H +unix` which disables the implicit addition of `-I +unix` breaking compilation (again, because `Unix` __is__ a direct dependency of these libraries).

- The third commit contains some tests. Writing the tests was a bit involved due to the need to have them pass both with versions of OCaml supporting `-H` and those that do not. Suggestions for improvement welcome.

- The last commit bumps the Dune version to 3.17. We could consider not bumping the Dune language version since the new mode is almost always a strict improvement for users of `(implicit_transitive_deps false)` (modulo corner cases such as the one mentioned in the second bullet point). Opinions welcome.

Future work:

- Adapt the Merlin machinery to pass `-H` as needed (there is already an open PR for this: #10535).
- Maybe adapt the `ocaml-index` machinery accordingly (cc @voodoos)

Fixes #9333 
Fixes ocsigen/tyxml#274 
Fixes #2733
Fixes #4963

### CODE REVIEW

**Comment by emillon** [comment]
Hi, thanks for this PR and the detailed description. There are a couple things we can land first:
- the 3.17 language version (we'll need it anyway and 3.16 has branched)
- the fix for unix is useful anyway because it triggers a warning on 5.x and we're going to bump the test suite to use this version soon


Regarding your question:
> Writing the tests was a bit involved due to the need to have them pass both with versions of OCaml supporting -H and those that do not. Suggestions for improvement welcome.

Yes. the reason it's painful is that we're only testing a single version of the compiler in CI, but you can add tests that require 5.x and mark them with `(enabled_if (>= %{ocaml_version} 5.2))`. This means that `make test` will work for them locally, and will work on 5.x in CI when we bump the CI version. Similarly for tests that rely on 4.x. They'll become dead tests when we bump the tested version but that can still be useful for some time.

**Comment by MA0010** [comment]
Hello, thanks for the comment.

> -   the 3.17 language version (we'll need it anyway and 3.16 has branched)
> - the fix for unix is useful anyway because it triggers a warning on 5.x and we're going to bump the test suite to use this version soon.

I have opened a separate PR to land the changes you proposed (see #10645). As you mentioned, those can be merged independently of the feature to be added through this PR. I will remove them from this PR as soon as they get merged successfully.

File: src/dune_rules/lib.mli
Comment (rgrinberg) [inline src/dune_rules/lib.mli:239]: I don't think you need this since our `Table` module is polymorphic. `let table = Table.create (module Lib)` should work

File: src/dune_rules/lib_flags.mli
Comment (rgrinberg) [inline src/dune_rules/lib_flags.mli:30]: Why can't we just pass a list: `([ Direct | Indirect] * Lib.t) list` as an argument? I'm not a fan of passing around callbacks.

File: src/dune_rules/compilation_context.ml
Comment (rgrinberg) [inline src/dune_rules/compilation_context.ml:151]: I think this mode still needs to be guarded behind `implicit_transitive_deps`. Once we experiment with it more, we can enable it by default for newer compilers perhaps.

**Comment by rgrinberg** [comment]
@voodoos could you review the implementation as well? I think you needed something similar for merlin. It would be good to see what can be shared.

File: src/dune_rules/compilation_context.ml
Comment (MA0010) [reply src/dune_rules/compilation_context.ml:151]: Sorry, but I do not seem to have understood well your point. The new feature is invoked only when `implicit_transitive_deps` is false, according to the written logic, or am I missing something?

Thanks for the review.

**Review by voodoos — COMMENTED** [review]
I think the way dependencies are exposed right now is too restricting. The `Lib.Compile.t` API only provide us with either `require_compile`, the list of direct dependencies, or  `require_link`, the list of all transitive dependencies  including the direct ones.

This means that we have to compute the disjunction `require_link \ require_compile` to get the actual "hidden" or "extra" dependencies. Right now this is needed in 3 places: rules for indexation (#10422), rules for merlin config (#10535) and this PR.

One issue is that `require_compile` and `require_link` are lists which prevent an efficient computation of the disjunction. In #10422 and #10535 this is done in a somewhat inefficient way using multiple calls to `List.mem`. This PR introduces a more efficient approach by using a set (the (lib, unit) Hashtbl.t) to perform more efficient membership tests.

I think we should update `Lib.Compile.t` to either provide an additional `requires_extra` list of libs that is computed in an efficient way and share the result, or have `requires_compile` and `requires_link` be Sets so that consumers can efficiently compute the disjunction without too much code duplication

This would allow the different places using the "extra deps" to do it in a uniform way.

File: src/dune_rules/lib_flags.mli
Comment (MA0010) [reply src/dune_rules/lib_flags.mli]: Hello,
I have removed the passed callback. Instead, now two lists are passed, one containing the direct dependencies (`requires_direct`) and the other contains the hidden dependencies (`requires_hidden`). The `requires_hidden` is also exposed in `Compilation_context.t`, and so can be used in other places when needed. 

The code have changed a little bit, and the diff in `lib_flags.ml` is now smaller.

Review appreciated!

**Comment by MA0010** [comment]
Concerning the last two commits, they contain edits based on the reviews and comments (cc @voodoos @rgrinberg), resting the main goal of keeping track of hidden deps unchanged. Those dependencies are now tracked by a new list `hidden_requires` exposed in the `Compilation_context.t` type. In short:

- The first of the new commits keeps the previous approach of using hashtables to compute the hidden dependencies, but now instead of being passed by a callback, they are stored in list `requires_hidden` and exposed in `Compilation_context.t` for external use cc (@voodoos). This list is then used to insert the `-H` flag for the correct deps.

- The second one simply removes the custom hashtable module from `lib.ml`, in favor of the already defined `Table` module (@rgrinberg). Also, the `ocaml_index.ml` file is adapted to use the new `requires_hidden` list, however in a naive way, and maybe can be used to adapt the machinery better (cc @voodoos )

File: src/dune_rules/lib_flags.ml
Comment (rgrinberg) [inline src/dune_rules/lib_flags.ml:174]: Isn't this re-ordering the include flags? e.g. if we have `(libraries x y)`, we'd have `closure(x) closure(y)`, and now we have `x y closure(x) closure(y)`?

Comment (nojb) [reply src/dune_rules/lib_flags.ml:174]: If I understand things correctly, I don't think anything is being re-ordered; instead some extra `-H` flags are being added at the "end" of the command-line when `implicit_transitive_deps` is `false`:

- If `implicit_transitive_deps` is `true`, then `direct_includes` is `requires_link` (from `Compilation_context`), and `hidden_includes` is empty; this is the same as today, and nothing changes.
- If `implicit_transitive_deps` is `false` (and OCaml is recent enough, etc) `direct_includes` is `requires_compile` (from `Compilation_context`), and `hidden_includes` contains some extra flags relative to today, but the order of `direct_includes` does not change.

Do you agree with this understanding @MA0010?

File: src/dune_rules/lib_flags.ml
Comment (nojb) [inline src/dune_rules/lib_flags.ml:173]: Let's share the code for the two definitions.

File: src/dune_rules/lib_flags.ml
Comment (nojb) [inline src/dune_rules/lib_flags.ml:101]: Let's share the code with `to_iflags` above.

File: src/dune_rules/compilation_context.ml
Comment (nojb) [inline src/dune_rules/compilation_context.ml:10]:
```suggestion
    ~hidden_requires
```

File: src/dune_rules/lib_flags.ml
Comment (Leonidas-from-XIV) [inline src/dune_rules/lib_flags.ml:88]: In our codebase we often try to avoid default flags, as the defaults are often easy to miss and in this case "to_iflags" which generates `-H` flags is somewhat strange.

As such I suggest making this a labelled `to_flag` function and 
```ocaml
let to_ifags = to_flags "-I"
let to_hflags = to_flags "-H"
```

This prevents having to change all the code while allowing an easy way to get an `-H` flag.

Comment (Leonidas-from-XIV) [reply src/dune_rules/lib_flags.ml]: You might not even need to expose `to_hflags` in the `mli` if nobody outside the module is going to use it. That way the compiler can warn about unused functions in case the code stops using `to_hflags`.

Comment (MA0010) [reply src/dune_rules/lib_flags.ml]: You are right for now `to_hflags` is not used anywhere outside `lib_flags.mli`, but I exposed it for the sake of symmetry (since if `to_iflags` is to be used somewhere in some later work, `to_hflags` would be also needed).

But maybe it is better for now to remove it from the interface.

File: src/dune_rules/compilation_context.ml
Comment (nojb) [inline src/dune_rules/compilation_context.ml:147]: Is this change needed?

File: src/dune_rules/compilation_context.ml
Comment (nojb) [inline src/dune_rules/compilation_context.ml:166]: `500` seems too large; I suggest `5`.

File: doc/changes/10644.md
Comment (nojb) [inline doc/changes/10644.md:6]:
```suggestion
- Add support for the -H flag (introduced in OCaml compiler 5.2) in dune
  (only supported for lang versions 3.17 and higher). This adaptation gives
  the correct semantics for `implicit_transitive_deps = false` mode.
  (#10644, fixes #9333, ocsigen/tyxml#274, #2733, #4963, @MA0100)
```

File: src/dune_rules/merlin/ocaml_index.ml
Comment (nojb) [inline src/dune_rules/merlin/ocaml_index.ml:35]: This doesn't seem quite right. For new versions of OCaml (>= 5.2) and version of the Dune language >= 3.17, Dune will pass `-H` to the compiler and these flags will be recorded in the `.cmt` files and will, in turn, be read by `ocaml-index`. Accordingly, in these cases we do not need to do anything special here. It is only for backwards compatibility that we can keep the above logic (ie if OCaml is < 5.2 _or_ if the Dune language veresion if < 3.17). Also, the flags should all be passed with `-I`, not `-H` in this case.

Comment (nojb) [reply src/dune_rules/merlin/ocaml_index.ml:35]: @voodoos sounds OK?

**Comment by nojb** [comment]
Friendly ping @anmonteiro: are you planning to review the PR? I feel this is getting ready to be merged, so do not hesitate to leave a note if you are planning to do so (which would be appreciated, of course!).

File: src/dune_rules/merlin/ocaml_index.ml
Comment (voodoos) [reply src/dune_rules/merlin/ocaml_index.ml:35]: Yes, that's completely right. If Dune >= 3.17 there is no more additional flags to pass to the indexer. (Note that indexing is not available prior to OCaml 5.2, so there is no need to check that here.)

Comment (nojb) [reply src/dune_rules/merlin/ocaml_index.ml:35]: Thanks for the confirmation! So only the check for OCaml lang version < 3.17 is needed.

Comment (MA0010) [reply src/dune_rules/merlin/ocaml_index.ml]: Done!

File: src/dune_rules/merlin/ocaml_index.ml
Comment (nojb) [inline src/dune_rules/merlin/ocaml_index.ml:41]: I suggest tweaking the placement and wording of the comment as follows:
```suggestion
      let scope = Compilation_context.scope cctx in
      (* Dune language >= 3.17 correctly passes the `-H` flag to the compiler. *)
      if Dune_project.dune_version (Scope.project scope) < (3, 17)
      then
        let open Resolve.Memo.O in
        let+ non_compile_libs =
          let* req_link = Compilation_context.requires_link cctx in
          let+ req_compile = Compilation_context.requires_compile cctx in
          List.filter req_link ~f:(fun l ->
            not (List.exists req_compile ~f:(Lib.equal l)))
        in
        Lib_flags.L.include_flags non_compile_libs [] (Lib_mode.Ocaml Byte)
      else Resolve.Memo.return Command.Args.empty
```

File: doc/changes/10644.md
Comment (nojb) [inline doc/changes/10644.md:2]:
```suggestion
  (requires lang version 3.17). This adaptation gives
```

File: src/dune_rules/compilation_context.ml
Comment (nojb) [inline src/dune_rules/compilation_context.ml:167]:
```suggestion
        let+ requires_compile = requires_compile
        and+ requires_link = Memo.Lazy.force requires_link in
        let requires_table = Table.create (module Lib) 5 in
        let () = List.iter ~f:(fun lib -> Table.set requires_table lib ()) requires_compile in
        List.filter requires_link ~f:(fun l -> not (Table.mem requires_table l))
```

**Review by nojb — COMMENTED** [review]
LGTM, modulo a couple of suggestions.

I would prefer a second official approval before signing this off.

File: src/dune_rules/compilation_context.ml
Comment (emillon) [inline src/dune_rules/compilation_context.ml:35]:
```suggestion
                       (direct_libs @ hidden_libs)
```

File: src/dune_rules/compilation_context.ml
Comment (emillon) [inline src/dune_rules/compilation_context.ml:44]:
```suggestion
                       (direct_libs @ hidden_libs)
```

File: src/dune_rules/lib_flags.mli
Comment (emillon) [inline src/dune_rules/lib_flags.mli:27]: Since it takes 2 arguments of type `t`, I think it'd be useful to label these.

File: test/blackbox-tests/test-cases/hidden-deps.t/bar.ml
Comment (emillon) [inline test/blackbox-tests/test-cases/hidden-deps.t/bar.ml:3]: the newline at eof in this test (and others) need fixing

File: test/blackbox-tests/test-cases/hidden-deps.t/run.t
Comment (emillon) [inline test/blackbox-tests/test-cases/hidden-deps.t/run.t:7]: I would recommend doing that in 2 tests:
- one that assumes ocaml >= 5.2 (guarded with `(enabled_if (>= %{ocaml_version} 5.2))`)
- one that assumes ocaml < 5.2 (similarly guarded)

And then you can use "normal" cram tests, which would consist in just running `getincludes` and displaying the output.

File: test/blackbox-tests/test-cases/hidden-deps.t/tyxml/run.t
Comment (emillon) [inline test/blackbox-tests/test-cases/hidden-deps.t/tyxml/run.t:4]: same here, you can guard the test behind `(enabled_if)`.

**Review by emillon — COMMENTED** [review]
Thanks. It seems you're mostly there so I left comments on style and tests. You might also want to add a quick note in `implicit_transitive_deps.rst` to explain that on ocaml 5.2 it's implemented using `-H`.

File: src/dune_rules/compilation_context.ml
Comment (rgrinberg) [reply src/dune_rules/compilation_context.ml:159]: Yeah, that wasn't very clear at all. Let me explain the behavior I would like to achieve:

When we're using `(implicit_transitive_deps false)` and OCaml >= 5.2.0, we should use use your new implementation with the `-H` flag.

When we're using `(implicit_transitive_deps false)` and OCaml < 5.2.0, we should go back to using the old implementation of hiding `-I` flags.

When we're using `(implicit_transitive_deps true)`, it should disable using `-H` flags completely.

The issue with the current behavior is that users do expect upgrading to 3.17 to be relatively easy. With this change, they will now need to look at all their transitive deps. While I agree that it's a good thing, it's too much of a breakage to introduce in a minor version bump.

Therefore, I suggest that we guard this behind a feature flag. We already have a feature flag though `implicit_transitive_deps`, so we can just reuse it.

**Review by rgrinberg — CHANGES_REQUESTED** [review]
As stated, I think this change is a bit too aggressive to be enabled by default at the moment.

**Comment by nojb** [comment]
> Let me explain the behavior I would like to achieve:

@rgrinberg Unless I am missing something, I think this is already the implemented behaviour: `-H` is only used when `implicit_transtiive_deps` is `false` (and OCaml >= 5.2); see condition on line 156 in the file in question:

https://github.com/ocaml/dune/blob/8e1757ffba6f5fff3584fbda94a3e4a7f478bf60/src/dune_rules/compilation_context.ml#L156-L157

File: test/blackbox-tests/test-cases/hidden-deps.t/run.t
Comment (MA0010) [reply test/blackbox-tests/test-cases/hidden-deps.t/run.t:7]: Hello, thanks for the review.
I am familiar with gaurding executables and libraries with the `enabled_if` (essentially making them hidden to the test). However, this would not be very useful here.

I have tried to add the `enabled_if` to a `cram` stanza in the dune file, hoping that would disable the whole test. This seems to not work since when I try to build the test it still runs irrespective of the condition mentioned with `enabled_if`.

Do you have any suggestions on how one would use `enabled_if` in order to disable the whole test?

Comment (emillon) [reply test/blackbox-tests/test-cases/hidden-deps.t/run.t:7]: sure! you can have a look at this for example:

https://github.com/ocaml/dune/blob/3.16.0/test/blackbox-tests/test-cases/dune#L103-L107

**Comment by MA0010** [comment]
I have completed the suggested edits. However, I request a final look at the tests before merging. For now, I have separated the `hidden-deps` tests into two tests, each one guarded by the correct OCaml version. However, this causes code duplication, and there may be a way to guard individual components of the same test behind the OCaml version.

File: src/dune_rules/compilation_context.ml
Comment (nojb) [reply src/dune_rules/compilation_context.ml:41]: @MA0010 this can be committed, right?

File: src/dune_rules/compilation_context.ml
Comment (nojb) [inline src/dune_rules/compilation_context.ml:165]:
```suggestion
        List.iter ~f:(fun lib -> Table.set requires_table lib ()) requires_compile;
```

File: test/blackbox-tests/test-cases/hidden-deps-supported.t/tyxml/run.ml
Comment (nojb) [inline test/blackbox-tests/test-cases/hidden-deps-supported.t/tyxml/run.ml:2]: Missing newline at EOF.

File: test/blackbox-tests/test-cases/hidden-deps-supported.t/tyxml/dune
Comment (nojb) [inline test/blackbox-tests/test-cases/hidden-deps-supported.t/tyxml/dune:3]: Missing newline at EOF.

File: test/blackbox-tests/test-cases/hidden-deps-supported.t/run.ml
Comment (nojb) [inline test/blackbox-tests/test-cases/hidden-deps-supported.t/run.ml:1]: Missing newline at EOF.

File: test/blackbox-tests/test-cases/hidden-deps-supported.t/run.t
Comment (nojb) [inline test/blackbox-tests/test-cases/hidden-deps-supported.t/run.t:24]:
```suggestion
```

File: test/blackbox-tests/test-cases/hidden-deps-supported.t/run.t
Comment (nojb) [inline test/blackbox-tests/test-cases/hidden-deps-supported.t/run.t:14]:
```suggestion
```

File: test/blackbox-tests/test-cases/hidden-deps-supported.t/run.t
Comment (nojb) [inline test/blackbox-tests/test-cases/hidden-deps-supported.t/run.t:44]:
```suggestion
```

File: test/blackbox-tests/test-cases/hidden-deps-unsupported.t/run.t
Comment (nojb) [inline test/blackbox-tests/test-cases/hidden-deps-unsupported.t/run.t:25]:
```suggestion

```

File: test/blackbox-tests/test-cases/hidden-deps-unsupported.t/run.t
Comment (nojb) [inline test/blackbox-tests/test-cases/hidden-deps-unsupported.t/run.t:15]:
```suggestion
```

**Comment by nojb** [comment]
I think all issues have been addressed and the PR is ready to go.

@emillon: would you like to do the honours? (as I have been supervising the PR, I prefer not to merge it myself).

**Comment by nojb** [comment]
> I think all issues have been addressed and the PR is ready to go.
> 
> @emillon: would you like to do the honours? (as I have been supervising the PR, I prefer not to merge it myself).

Tha PR should be squash-merged by the way.

**Comment by emillon** [comment]
@nojb yes, I'll give it a last look, squash and merge.

File: doc/changes/10644.md
Comment (emillon) [inline doc/changes/10644.md:2]:
```suggestion
  (requires dune lang 3.17). This adaptation gives
```

**Comment by emillon** [comment]
I can't push to the original branch but that should be fine.

**Comment by nojb** [comment]
> I can't push to the original branch but that should be fine.

I think this was my fault (I opened the PR from a branch in @MA0010's fork which disables this mechanism).

@MA0010 can you fix the remaining issue so that we can merge? Thanks!

**Comment by emillon** [comment]
the last issues are just cosmetic - newlines etc and the changelog can be fixed at release time.
Thanks a lot to everyone involved here. This is the consequence of a multi-year effort that started in dune 1.7 (5 years ago), made its way to the compiler, and back here!

**Comment by kit-ty-kate** [comment]
Thank you so much for this work!

**Comment by Alizter** [comment]
Is tyxml supposed to be a test dependency here?

**Comment by nojb** [comment]
> Is tyxml supposed to be a test dependency here?

Yes, there is a test that depends on it, so I guess so.

**Comment by rgrinberg** [comment]
I think we should modify the test to avoid this dependency

**Comment by nojb** [comment]
> I think we should modify the test to avoid this dependency

This specific test was added simply to exhibit the resolution of https://github.com/ocsigen/tyxml/issues/274. But there are other tests for `-H`, so we can just remove this one test.

**Comment by nojb** [comment]
See #11728

**Pull request merged**
